Ptc 161, project selector

exi opened the pull request:
This does not include a 'themed' version, just the base code

Commit:
Move the buttons below the header, just above the table/graph

This emphasizes that they belong to the graph only.

Commit:
fix small layouting bug

put the buttons in the right place to make the layout consistent

Commit:
preselect last booked project in project selector

Commit:
move project chooser into separate widget

Commit:
cleanup dependencies

Commit:
reformat dependency list of booking creator

just make it one per line

Commit:
add dummy projekt chooser to statistics page

Commit:
modify timecard model to accept optional projects filter

just add a projects parameter to some functions

Commit:
add projects filter to timecard index controller

make the timecard index controller projects filter aware.

Commit:
remove static worktime month graph from the statistics template

Commit:
replace project choose with container in statistics view

Commit:
make frontend timecard model projects filter aware

Commit:
make worktimemonthtable and worktimemonthgraph projects filter aware

Commit:
refactor projectchooser to allow custom option generation

create overridable functions "createOptions" and "postStoreSet" to allow custom
options.

Commit:
make statisticsView project selector aware

Commit:
add separation line projectChooser workarounds

we don't want the user to be able to select the separator lines

voidus commented:
Please don't name your branches by only the ticket id, something like projectselector-161 would be nicer

exi commented:
I thought giving it the github title 'Ptc 161, project selector' would be enough

Commit:
don't use getModelObject anymore

Commit:
prevent change event emitting before the projectChooser data has been loaded.
otherwise we have multiple change events that trigger reloads upstream

Commit:
make the monthgraph/monthtable load depend on the projectchooser change event.

we don't load the monthGraph right away, because it will be reloaded anyway
once the projectChooser starts up and fires it's 'change' event.
By just settings the monthState, the projectChoosers change event will
trigger a graph/table reload once it has finished starting up.
This prevents a double initialization of the monthGraph which was sometimes
visible as a visual 'flickering' of the graph

Commit:
add missing destroy checks in monthgraph and table

Commit:
remove unnecessary dom emptying from monthgraph.

this change was not intentional, just a line i added by accident and forgot to
remove it.

Commit:
rename _onProjectChange function to _updateMonthWidget

Commit:
factor out projectChooser for statistics to enhance readability

Commit:
change default value to null in timecard model project filter

Commit:
Merge remote-tracking branch 'mainline/next' into PTC-161

Conflicts:
	phprojekt/htdocs/phpr/BookingList.js
	phprojekt/htdocs/phpr/BookingList/BookingCreator.js
	phprojekt/htdocs/phpr/StatisticsView.js
	phprojekt/htdocs/phpr/template/statisticsView.html

Commit:
Merge remote-tracking branch 'mainline/next' into PTC-161

Commit:
make getMonthRecords static

we don't want to use getModelObject() anymore and the function does not hold
locale state anyway

Commit:
fix _projectsParamToArray helper to use null as the default
